### PR TITLE
Fix int16 overflow in the gather_qmm sorted row bound

### LINKED ISSUE #3856: Silent numerical corruption in single long forward on quantized MoE when sequence length % 32 != 0 (Qwen3-Coder-30B-A3B-8bit, M5)
## Summary

A single `model(ids)` forward over a long prompt on a quantized MoE model
produces silently corrupted activations/KV whenever the forward's sequence
length is **not a multiple of 32**. Feeding the identical token ids through
the identical model in 2048-token chunks (what `mlx_lm.generate_step` does)
gives a different — and behaviorally correct — result. The divergence is
deterministic, content-independent (reproduces with random token ids), large
(fp16 K-cache entries off by up to ~86; final-position logits off by ~17–23;
greedy argmax flips), and big enough to break downstream behavior: a
needle-in-haystack QA that scores 100% recall with chunked prefill scores 0%
with a one-shot prefill at N≈16k.

This looks like a sibling of the 0.31.x quantized-matmul alignment bugs
(`gather_qmm` >32k-row MoE corruption; unaligned-N `quantized_matmul` with
m ≥ 10, m % 32 ≠ 0 — both fixed in 0.32.0). The MoE expert path appears to
still have an unaligned-M case: dense models (Qwen3-8B/32B, same quant
recipes, same lengths) are behaviorally clean in the same A/B, and the first
divergent tensor we observe is layer 1's K — i.e. the corruption enters
through layer 0's block output. Since dense models exercise the same
attention path at these shapes cleanly, the MoE FFN (`gather_qmm` /
gather-QMM-family kernels) is the prime suspect.

## Environment

- mlx **0.32.0.dev20260708+1ac2fbac** and **0.32.1.dev20260715+cf3ce39c**
  (both reproduce; source builds with NAX kernels enabled,
  `MACOSX_DEPLOYMENT_TARGET=26.2`)
- Apple M5, macOS 26 (Darwin 25.5.0), Python 3.12
- Model: `mlx-community/Qwen3-Coder-30B-A3B-Instruct-8bit` (affine 8-bit,
  MoE 30B-A3B). Dense `mlx-community/Qwen3-8B-4bit` / `Qwen3-32B-4bit` do
  **not** reproduce behaviorally at the same lengths.
- mlx-lm used only to load the model and build the KV cache.

## Repro

```python
# python repro.py            -> single length
# python repro.py --bisect   -> alignment scan
import argparse
import mlx.core as mx
from mlx_lm import load
from mlx_lm.models.cache import make_prompt_cache

def cache_diff(model, ids, chunk=2048):
    def build(step):
        cache = make_prompt_cache(model)
        for s in range(0, len(ids), step):
            logits = model(mx.array([ids[s:s+step]]), cache=cache)
            mx.eval(logits, [c.state for c in cache])
        return cache, logits[0, -1]
    c1, l1 = build(len(ids))   # one-shot
    c2, l2 = build(chunk)      # chunked
    worst_k = 0.0
    for a, b in zip(c1, c2):
        if getattr(a, "keys", None) is None:
            continue
        n = min(a.offset, b.offset)
        dk = float(mx.max(mx.abs(a.state[0][..., :n, :] - b.state[0][..., :n, :])).item())
        worst_k = max(worst_k, dk)
    dl = float(mx.max(mx.abs(l1 - l2)).item())
    same = int(mx.argmax(l1).item()) == int(mx.argmax(l2).item())
    return worst_k, dl, same

ap = argparse.ArgumentParser()
ap.add_argument("--model", default="mlx-community/Qwen3-Coder-30B-A3B-Instruct-8bit")
ap.add_argument("--bisect", action="store_true")
args = ap.parse_args()

model, tok = load(args.model)
mx.random.seed(0)
ids = mx.random.randint(0, 50000, (17000,)).tolist()
lengths = [8192, 12288, 16000, 16032, 16064, 16065, 16066, 16067, 16068,
           16128, 16384, 16900] if args.bisect else [16068]
print(f"mlx {mx.__version__}")
print(f"{'N':>7} {'worstK':>8} {'logitdiff':>9}  argmax")
for L in lengths:
    wk, dl, ok = cache_diff(model, ids[:L], 2048)
    print(f"{L:>7} {wk:>8.3f} {dl:>9.3f}  {'OK' if ok else 'DIVERGED'}", flush=True)
```

## Observed (random ids, seed 0, deterministic across runs)

```
mlx 0.32.0.dev20260708+1ac2fbac        (0.32.1.dev20260715: same pattern)
      N   worstK logitdiff  argmax
   8192   15.125     0.562  OK          # 32-aligned
  12288   15.125     1.172  OK          # 32-aligned
  16000   15.125     0.182  OK          # 32-aligned
  16032   15.125     0.344  OK          # 32-aligned
  16064   26.000     0.109  OK          # 32-aligned
  16065   32.000     7.562  DIVERGED    # N % 32 == 1
  16066   32.500     8.312  DIVERGED    # N % 32 == 2
  16067   32.000     8.375  DIVERGED    # N % 32 == 3
  16068   32.000    17.500  DIVERGED    # N % 32 == 4
  16128   26.000     1.148  OK          # 32-aligned
  16384   26.000     0.188  OK          # 32-aligned
  16900   32.000     9.438  DIVERGED    # N % 32 == 4
```

Every diverging length is unaligned; every 32-aligned length matches argmax
with small logit deltas. Unaligned lengths below 16k already show large
numeric error without flipping the argmax at the final position (e.g. on
0.32.1.dev20260715: N=8193 → worstK 131, logitdiff 17.0, argmax OK; N=12289 →
worstK 54, logitdiff 4.3, argmax OK), so short-prompt spot checks miss it.

Notes:

- The constant worstK ≈ 15–26 at aligned lengths is presumably benign
  kernel-tiling numerics amplified by outlier channels (argmax stable, logit
  deltas ≪ 1 in most rows); the unaligned case is qualitatively different.
- Layer-0 K is identical between the two paths (it depends only on the
  embeddings); divergence first appears in layer-1 K and compounds through
  all 48 layers.
- Real-workload impact: with a real 16,068-token needle-in-haystack prompt,
  the one-shot path's K diverges by up to 86, final logits by 23.4, and
  greedy QA recall flips 1.0 → 0.0 (the model emits repetitive
  off-task text). Chunked prefill of the same ids answers correctly. This
  cost us a multi-day false negative in a KV-eviction study before we traced
  it to the prefill path.

## Workaround

Chunk all prefills (any chunk size that keeps each forward 32-aligned or
short works; we use 2048).

We can test candidate fixes, run finer-grained bisects, or collect kernel-level
traces on M5 hardware on request.

### PR DESCRIPTION
**Pull request by erwinzhang7**
Fixes #3856.

A quantized MoE silently returns wrong results for a single long prefill on M5. The same tokens fed through in chunks are correct, so the corruption is invisible unless you compare the two.

This is independent of #4009. I verified that by running the issue's repro on 0.32.0 and on a build carrying #4009, and the outputs were identical, so the two defects do not interact.

## Root cause

In `affine_gather_qmm_rhs_nax`:

```cpp
const short sgp_sm = align_M ? SM : min(SM, short(max(0, (M - (y_row + tm)))));
const short sgp_sn = align_N ? SN : min(SN, short(max(0, (N - (y_col + tn)))));
```

The narrowing to `short` happens inside the `min`. Once `M` passes the int16 range the inner expression wraps negative, the `min` returns that negative value, and the tile is then loaded with a garbage row bound. Taking the `min` in `int` and narrowing after is safe, since the result is bounded by `SM`.

Two details make this hard to hit by accident:

1. The unaligned branch only runs when `align_M` is false, so `M % BM != 0` is required as well as `M` being large.
2. The line directly above it already does it in the safe order, which is why nothing else in the kernel misbehaves:

```cpp
const short tgp_bm = align_M ? BM : short(min(BM, M - y_row));
```

`fp_quantized_nax.h:864` and every other `sgp_sm` site in the tree already use the `min(int(SM), ...)` form. These two lines were the only outliers.

## Trigger condition, and a correction to the issue title

The issue reports the trigger as sequence length `% 32 != 0`. The actual condition is the kernel row count `M > 32767` **and** `M % 64 != 0`.

In `SwitchGLU` the rows are sorted and flattened, so `M = tokens * experts_per_token`. Qwen3-Coder-30B-A3B routes to 8 experts per token, giving `M = 8L`, so `M % 64 != 0` reduces to `L % 8 != 0`. The lengths sampled in the issue (16000, 16032, 16064 clean; 16065, 16066 corrupt) happen to satisfy both `% 32` and `% 8`, so they cannot distinguish the two rules. `L % 8` is the correct one.

## Evidence

Synthetic, no model required. `gather_qmm` with `sorted_indices=True` against `sorted_indices=False` on the same already sorted input, E=4, N=256, K=512, group_size 64, 8 bit. K is aligned here so #4009 is not involved. Max difference relative to `mean(abs(unsorted))`:

| M | M % 64 | M > 32767 | before | after |
| --- | --- | --- | --- | --- |
| 4000 | 32 | no | 0.0035 | 0.0035 |
| 32704 | 0 | no | 0.0035 | 0.0035 |
| 32760 | 56 | no | 0.0035 | 0.0035 |
| 32768 | 0 | yes | 0.0035 | 0.0035 |
| **32800** | **32** | **yes** | **4.894** | **0.0035** |
| 32832 | 0 | yes | 0.0035 | 0.0035 |
| 40000 | 0 | yes | 0.0035 | 0.0035 |
| **40001** | **1** | **yes** | **6.517** | **0.0035** |
| 65600 | 0 | yes | 0.0035 | 0.0035 |

Every corrupt case satisfies both conditions and no clean case satisfies both.

End to end on Qwen3-Coder-30B-A3B-Instruct 8 bit, M5 Max, comparing a single prefill against 2048 token chunks. Max absolute difference in the final position logits:

| L | L % 8 | before | after |
| --- | --- | --- | --- |
| 8192 | 0 | 0.156 | 0.156 |
| 12288 | 0 | 0.352 | 0.352 |
| 16000 | 0 | 0.367 | 0.367 |
| 16032 | 0 | 0.250 | 0.250 |
| 16064 | 0 | 0.281 | 0.281 |
| **16065** | **1** | **9.875** | **0.156** |
| **16066** | **2** | **8.875** | **0.227** |

Worst KV cache difference at the two corrupt lengths drops from 43.75 and 43.0 to 20.0, matching the aligned baseline.

One note on reading that table: a residual argmax flip remains at L=16000, unchanged by this patch. It is not corruption. At that position the top two logits are exactly tied in fp16 (both 9.6250, gap 0.0000), so which one wins is decided by ordinary accumulation noise between the one shot and chunked paths. Every length now sits in the same 0.15 to 0.37 band.

## Test

Added `test_gather_qmm_sorted_large_m`, covering M of 32704 (under the int16 range, as a control), 32800 and 40001. It fails on 0.32.0 on the latter two and passes with this change.

`test_quantized.py`, `test_blas.py`, `test_nn.py` and `test_ops.py` pass, 279 tests.

## Performance

No measurable change. The edit only widens a scalar min computed once per thread, and the difference is well inside the run to run variance of a `gather_qmm` benchmark.

### CODE REVIEW

**Comment by zcbenz** [comment]
Closing since there are already PRs with similar changes.
